### PR TITLE
Implement the DynamicLookupProtocol proposal.

### PR DESCRIPTION
**Pull request by lattner**
This is an implementation of the DynamicLookupProtocol that I am proposing for inclusion in Swift.  It has no known implementation problems, but I absolutely welcome feedback on the implementation approach.

The approach is straight-forward.  It does not bloat any core data structures and is not invasive into the compiler.  High points of the patch:
- Add the new protocol to the standard library.
- When ConstraintSystem::performMemberLookup is about to fail a value member lookup, check to see if the base type conforms to DynamicMemberLookupProtocol.  If so, add the subscript(dynamicMember:) implementations to the candidate set as a new OverloadChoice::DynamicMemberLookup kind.
- CSApply handles OverloadChoiceKind::DynamicMemberLookup by calling buildSubscript().
- Add a test case handling many weird cases I could think of.

### CODE REVIEW

**Comment by lattner** [comment]
@swift-ci please test

**Comment by lattner** [comment]
I'll be out of town for the next few days, but I'll take a look at the test failures and fix them when I get back.

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -1287,12 +1287,14 @@ namespace {
                          ArrayRef<Identifier> argLabels,
                          bool hasTrailingClosure,
                          ConstraintLocatorBuilder locator,
-                         bool isImplicit, AccessSemantics semantics) {
+                         bool isImplicit, AccessSemantics semantics,
+                         Optional<SelectedOverload> selected = None) {
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: is this ever going to be None?

File: lib/Sema/CSApply.cpp [hunk 2]
```diff
@@ -1287,12 +1287,14 @@ namespace {
...
       // Determine the declaration selected for this subscript operation.
-      auto selected = getOverloadChoiceIfAvailable(
-                        cs.getConstraintLocator(
-                          locator.withPathElement(
-                            ConstraintLocator::SubscriptMember)));
+      if (!selected)
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: Maybe the caller should do this?

File: lib/Sema/CSApply.cpp [hunk 3]
```diff
@@ -1421,30 +1423,47 @@ namespace {
...
-                                  argLabels, hasTrailingClosure,
-                                  locator.withPathElement(
-                                    ConstraintLocator::SubscriptIndex));
-      if (!index)
-        return nullptr;
+      // Figure out the index and result types.
+      Type resultTy;
+      if (!isDynamicMemberLookup) {
+        auto subscriptTy = simplifyType(selected->openedType);
+        AnyFunctionType *subscriptFnTy = subscriptTy->castTo<AnyFunctionType>();
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: It will actually always be a FunctionType. GenericFunctionType cannot appear here

File: lib/Sema/CSApply.cpp [hunk 4]
```diff
@@ -2630,23 +2649,48 @@ namespace {
...
+
+        // Build and type check the string literal index value to the specific
+        // string type expected by the subscript.
+        Expr *nameExpr = new (cs.getASTContext())
+          StringLiteralExpr(selected.choice.getName().getBaseIdentifier().str(),
+                            nameLoc.getStartLoc(), /*implicit*/true);
+        auto nameLabel = cs.getASTContext().getIdentifier("dynamicMember");
+        
+        auto stringType = subscriptDecl->getIndices()->get(0)->getTypeLoc();
+        (void)cs.TC.typeCheckExpression(nameExpr, dc, stringType,
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: What if this fails?

File: lib/Sema/CSApply.cpp [hunk 5]
```diff
@@ -6417,7 +6461,8 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
   }
 
   // Unresolved types come up in diagnostics for lvalue and inout types.
-  if (fromType->hasUnresolvedType() || toType->hasUnresolvedType())
+  if (fromType->hasUnresolvedType() || toType->hasUnresolvedType() ||
+      fromType->is<ErrorType>())
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: Is there a test case that triggers this? You should split it out into a separate patch with a commit message explaining why

File: lib/Sema/CSApply.cpp [hunk 6]
```diff
@@ -7925,7 +7970,8 @@ Expr *TypeChecker::callWitness(Expr *base, DeclContext *dc,
 }
 
 Expr *
-Solution::convertBooleanTypeToBuiltinI1(Expr *expr, ConstraintLocator *locator) const {
+Solution::convertBooleanTypeToBuiltinI1(Expr *expr,
+                                        ConstraintLocator *locator) const {
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: It would be nicer if this and other formatting changes were in a separate patch

File: lib/Sema/CSSimplify.cpp [hunk 7]
```diff
@@ -2896,6 +2896,30 @@ getArgumentLabels(ConstraintSystem &cs, ConstraintLocatorBuilder locator) {
...
+    TC.Context.getProtocol(KnownProtocolKind::DynamicMemberLookupProtocol);
+
+  auto conformance =
+    TC.conformsToProtocol(ty, DMLP, DC, ConformanceCheckFlags::InExpression);
+  
+  // If this type doesn't conform to it, check to see if this is an existential
+  // that contains it.
+  if (!conformance)
+    conformance =
+      TC.containsProtocol(ty, DMLP, DC, ConformanceCheckFlags::InExpression);
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: You should check but I think containsProtocol() is a superset of conformsToProtocol() so you only need the former here

File: lib/Sema/CSSimplify.cpp [hunk 8]
```diff
@@ -2896,6 +2896,30 @@ getArgumentLabels(ConstraintSystem &cs, ConstraintLocatorBuilder locator) {
...
+  if (!conformance)
+    return false;
+  
+  // If we found something, make sure that it is a valid conformance.
+  return conformance->isAbstract() || !conformance->getConcrete()->isInvalid();
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: If ty is existential, the conformance will be abstract so you'll always return false here -- did you test the existential case?

File: lib/Sema/CSSimplify.cpp [hunk 9]
```diff
@@ -3286,6 +3310,40 @@ performMemberLookup(ConstraintKind constraintKind, DeclName memberName,
...
+  // that conforms to DynamicMemberLookupProtocol, then we resolve the reference
+  // to the subscript(dynamicMember:) member, and pass the member name as a
+  // string.
+  if (constraintKind == ConstraintKind::ValueMember && hasInstanceMembers &&
+      memberName.isSimpleName() && !memberName.isSpecial()) {
+    auto name = memberName.getBaseIdentifier();
+    if (containsOrConformsToDLMP(instanceTy, DC, name, TC)) {
+      // Recursively look up the subscript(dynamicMember:)'s in this type.
+      auto subscriptName =
+        DeclName(getASTContext(), DeclBaseName::createSubscript(),
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: A few places in this method use TC.Context and here you're using getASTContext(), might want to just extract a ctx local variable

File: lib/Sema/TypeCheckProtocol.cpp [hunk 10]
```diff
@@ -1748,6 +1748,81 @@ checkWitness(AccessScope requiredAccessScope,
...
 }
 
+/// Given a subscript defined as "subscript(dynamicMember:)->T", return true if
+/// it is an acceptable implementation of the DynamicMemberLookupProtocol
+/// requirement.
+bool swift::isAcceptableDynamicMemberLookupSubscript(ValueDecl *decl,
+                                                     DeclContext *DC,
+                                                     TypeChecker &TC) {
+  // Check the conforming decl.
+  auto match = cast<SubscriptDecl>(decl);
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Since you're casting unconditionally it makes more sense to just change the parameter type

File: lib/Sema/TypeCheckProtocol.cpp [hunk 11]
```diff
@@ -2262,7 +2337,14 @@ namespace {
       conformance->setInvalid();
       return conformance;
     }
-
+    
+    // If this is the DynamicMemberLookupProtocol, manually check its
+    // requirement.
+    if (Proto->isSpecificProtocol(
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: There are many other places where we directly call ModuleDecl::lookupConformance() and they will miss this check, but I'm not sure if it matters

**Comment by slavapestov** [comment]
I'm generally concerned with how long some of the functions like CSApply's buildSubscript() are getting. If you could split it up a bit it would be great

File: lib/Sema/CSSimplify.cpp [hunk 12]
```diff
@@ -2896,6 +2896,30 @@ getArgumentLabels(ConstraintSystem &cs, ConstraintLocatorBuilder locator) {
   return known->second;
 }
 
+
+// Return true if the specified type conforms to the DynamicLookupMemberProtocol
+// or if it is a protocol that inherits from it.
+static bool containsOrConformsToDLMP(Type ty, DeclContext *DC,
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: This function name doesn't read well to me. The important part is not "contains or conforms", it's "dynamic lookup member protocol". I would shorten the former and expand the latter

File: lib/Sema/CSSimplify.cpp [hunk 13]
```diff
@@ -3286,6 +3310,40 @@ performMemberLookup(ConstraintKind constraintKind, DeclName memberName,
       }
     }
   }
+  
+  // If we're about to fail lookup, but we are looking for members in a type
+  // that conforms to DynamicMemberLookupProtocol, then we resolve the reference
+  // to the subscript(dynamicMember:) member, and pass the member name as a
+  // string.
+  if (constraintKind == ConstraintKind::ValueMember && hasInstanceMembers &&
```
Comment (slavapestov) [inline lib/Sema/CSSimplify.cpp]: Maybe split out the body of this if statement into a new function, since performMemberLookup() is already quite long

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (lattner) [reply lib/Sema/CSApply.cpp]: Two of the three clients pass none here, so yep.

File: lib/Sema/CSApply.cpp (hunk 4 above)
Comment (lattner) [reply lib/Sema/CSApply.cpp]: This can't fail, because previous filtering has validated that the index type is ExpressibleByStringLiteral.  This is similar to the logic used forming string literals themselves.

File: lib/Sema/CSApply.cpp (hunk 6 above)
Comment (lattner) [reply lib/Sema/CSApply.cpp]: Indeed.  I split out the other general changes, but one off changes like this I left in.  They aren't systematic enough to cause problems for review I think, but I understand what you mean.

**Comment by lattner** [comment]
Thank you for the detailed feedback @slavapestov!  When I get a chance to come back to this later this week or early next week, I'll pull in changes for your feedback into the next version of the patch.  I appreciate it!

File: lib/Sema/CSApply.cpp (hunk 3 above)
Comment (lattner) [reply lib/Sema/CSApply.cpp]: Fixed, thanks!

File: lib/Sema/CSApply.cpp (hunk 5 above)
Comment (lattner) [reply lib/Sema/CSApply.cpp]: Removed, thanks.

File: lib/Sema/CSSimplify.cpp (hunk 12 above)
Comment (lattner) [reply lib/Sema/CSSimplify.cpp]: Good point, switched to isDynamicMemberLookupable

File: lib/Sema/CSSimplify.cpp (hunk 7 above)
Comment (lattner) [reply lib/Sema/CSSimplify.cpp]: Looks like it!  Thanks

File: lib/Sema/CSSimplify.cpp (hunk 8 above)
Comment (lattner) [reply lib/Sema/CSSimplify.cpp]: Yes, the tests include an existential case.

File: lib/Sema/CSSimplify.cpp (hunk 13 above)
Comment (lattner) [reply lib/Sema/CSSimplify.cpp]: performMemberLookup is long, but the logic is straight-forward.  I don't think that splitting it would make it easier to read or maintain, but it could always be done at some point independent of this patch.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 10 above)
Comment (lattner) [reply lib/Sema/TypeCheckProtocol.cpp]: Yes, good point, fixed.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 11 above)
Comment (lattner) [reply lib/Sema/TypeCheckProtocol.cpp]: I think it is fine.  Whatever defines a conformance will get checked.

**Comment by lattner** [comment]
I'm closing this, moving forward to PR 13361 which incorporates Slava's feedback and another change.  Thanks!
